File: haskell/scrabble-score/src/Scrabble.hs
module Scrabble (scoreLetter, scoreWord) where

import Data.Char (toUpper)

scores :: [(Integer, String)]
scores = [(1, "AEIOULNRST"),
           (2, "DG"),
           (3, "BCMP"),
           (4, "FHVWY"),
           (5, "K"),
           (8, "JX"),
           (10, "QZ")]

scoreLetter :: Char -> Integer
scoreLetter letter = head ((scores >>= checkMatch (toUpper letter)) ++ [0])
  where checkMatch ltr (pnt, ltrs) = [pnt | ltr `elem` ltrs]

scoreWord :: String -> Integer
scoreWord = sum . map scoreLetter

File: java/scrabble-score/src/main/java/Scrabble.java
import com.google.common.collect.ImmutableMap;

import static com.google.common.base.Strings.nullToEmpty;
import static java.util.Map.Entry;

public class Scrabble {
	private static final ImmutableMap<String, Integer> POINTS =
		ImmutableMap.<String, Integer>builder()
		.put("AEIOULNRST", 1)
		.put("DG", 2)
		.put("BCMP", 3)
		.put("FHVWY", 4)
		.put("K", 5)
		.put("JX", 8)
		.put("QZ", 10)
		.build();
	private static final int MISSING = -1;
	private static final int NO_POINTS = 0;

	private final String word;

	public Scrabble(final String word) {
		this.word = word;
	}

	public int getScore() {
		return nullToEmpty(word)
			.toUpperCase()
			.chars()
			.map(Scrabble::rateCharacter)
			.sum();
	}

	private static int rateCharacter(final int ch) {
		return POINTS.entrySet().stream()
			.filter(score -> score.getKey().indexOf(ch) != MISSING)
			.mapToInt(Entry::getValue)
			.findFirst()
			.orElse(NO_POINTS);
	}
}
